cmd/collectiondir/serve.go

```diff
@@ -0,0 +1,1012 @@
+		}
+	}
+}
+
+// handleFirehose consumes XRPCStreamEvent from firehoseThread(), further parses data and applies
+func (cs *collectionServer) handleFirehose(fhevents <-chan *events.XRPCStreamEvent) {
+	defer cs.wg.Done()
+	defer cs.log.Info("handleFirehose exit")
+	defer close(cs.ingestFirehose)
+	var lastSeq int64
+	lastSeqSet := false
+	notDone := true
+	for notDone {
+		select {
+		case <-cs.shutdown:
+			cs.log.Info("firehose handler shutdown")
+			notDone = false
+		case evt, ok := <-fhevents:
+			if !ok {
+				notDone = false
+				cs.log.Info("firehose handler closed")
+				break
+			}
+			firehoseReceivedCounter.Inc()
+			seq, ok := evt.GetSequence()
+			if ok {
+				lastSeq = seq
+				lastSeqSet = true
+			}
+			if evt.RepoCommit != nil {
+				firehoseCommits.Inc()
+				cs.handleCommit(evt.RepoCommit)
+			}
+		}
+	}
+	if lastSeqSet {
+		cs.pcd.SetSequence(lastSeq)
+	}
+}
+
+func (cs *collectionServer) handleCommit(commit *comatproto.SyncSubscribeRepos_Commit) {
+	for _, op := range commit.Ops {
+		// op.Path is collection/rkey
+		slash := strings.IndexRune(op.Path, '/')
+		if slash == -1 {
+			cs.log.Warn("bad op path", "repo", commit.Repo)
+			return
+		}
+		collection := op.Path[:slash]
+		firehoseCommitOps.WithLabelValues(op.Action).Inc()
+		if op.Action == "create" || op.Action == "update" {
+			firehoseDidcSet.Inc()
+			cs.ingestFirehose <- DidCollection{
+				Did:        commit.Repo,
+				Collection: collection,
+			}
+		}
+	}
+}
+
+func (cs *collectionServer) StartMetricsServer(ctx context.Context, addr string) error {
+	defer cs.wg.Done()
+	defer cs.log.Info("metrics server exit")
+	cs.metricsServer = &http.Server{
+		Addr:    addr,
+		Handler: promhttp.Handler(),
+	}
+	return cs.metricsServer.ListenAndServe()
+}
+
+func (cs *collectionServer) StartApiServer(ctx context.Context, addr string) error {
+	defer cs.wg.Done()
+	defer cs.log.Info("api server exit")
+	var lc net.ListenConfig
+	li, err := lc.Listen(ctx, "tcp", addr)
+	if err != nil {
+		return err
+	}
+	e := echo.New()
+	e.HideBanner = true
+
+	e.Use(MetricsMiddleware)
+	e.Use(middleware.CORSWithConfig(middleware.CORSConfig{
+		AllowOrigins: []string{"*"},
+		AllowHeaders: []string{echo.HeaderOrigin, echo.HeaderContentType, echo.HeaderAccept, echo.HeaderAuthorization},
+	}))
+
+	e.GET("/_health", cs.healthz)
+
+	e.GET("/v1/getDidsForCollection", cs.getDidsForCollection)
+	e.GET("/v1/listCollections", cs.listCollections)
+
+	// TODO: allow public 'requestCrawl' API?
+	//e.GET("/xrpc/com.atproto.sync.requestCrawl", cs.crawlPds)
+	//e.POST("/xrpc/com.atproto.sync.requestCrawl", cs.crawlPds)
+
+	// admin auth heador required
+	e.POST("/admin/pds/requestCrawl", cs.crawlPds) // same as relay
+	e.GET("/v1/crawlStatus", cs.crawlStatus)
+
+	e.Listener = li
+	srv := &http.Server{
+		Handler: e,
+	}
+	cs.apiServer = srv
+	return srv.Serve(li)
+}
+
+const statsCacheDuration = time.Second * 300
+
+type GetDidsForCollectionResponse struct {
+	Dids   []string `json:"dids"`
```

Comment:
returning "just a list of strings" is the simple thing here. I think we should consider doing an array of objects, each with the field `did` though. This is less efficient today, but we have found that we tend to want to add flags or other metadata in the future, and that is way easier if we do arrays-of-objects to start.

I would maybe lean towards "repos" as the top level key? "dids" reads weird.

(we can re-review and tweak API schema details in the `atproto` lexicon PR instead of here)

Reply:
"dids" is just a list of strings, "repos" invites further fields. could leave it as "dids" for now, and replace/extend it to "repos" objects later?